src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotStore.scala

```diff
@@ -166,49 +166,43 @@ class CassandraSnapshotStore(cfg: Config) extends SnapshotStore with CassandraSt
   }
 
   private def metadata(prepStmt: PreparedStatement, persistenceId: String, criteria: SnapshotSelectionCriteria): Future[Vector[SnapshotMetadata]] = {
-    // TODO the RowIterator is using some blocking, would benefit from a rewrite
-    Future {
-      new RowIterator(prepStmt, persistenceId, criteria.maxSequenceNr).map { row =>
-        SnapshotMetadata(row.getString("persistence_id"), row.getLong("sequence_nr"), row.getLong("timestamp"))
-      }.dropWhile(_.timestamp > criteria.maxTimestamp).toVector
-    }(blockingDispatcher)
-  }
+    def iterate(nr: Long, buf: ListBuffer[SnapshotMetadata]): Future[Vector[SnapshotMetadata]] = {
+      session.execute(prepStmt.bind(persistenceId, nr: JLong)) flatMap { rs =>
+        var rowCount = 0
+        var nextNr = 0l
+
+        def fetchAvailable(rs: ResultSet): Future[Vector[SnapshotMetadata]] = {
+          for (r <- rs.asScala.take(rs.getAvailableWithoutFetching)) {
+            nextNr = r.getLong("sequence_nr")
+            rowCount += 1
+            buf += SnapshotMetadata(r.getString("persistence_id"), nextNr, r.getLong("timestamp"))
+          }
+          if (rowCount < maxMetadataResultSize) Future.successful(buf.dropWhile(_.timestamp > criteria.maxTimestamp).toVector)
```

Comment:
Well, I tried to not change any logic while rewriting. If I put ">=" instead of "<" here, tests fail.